[laihuiqi] iP

laihuiqi opened the pull request:
# Chattime
> "Time flies, enjoy life."

Have a nice organized day start with an easy chat with our bot! It's :

- easy to use
- text-based 
- clever
- **FREE**
- ~~cute~~ SUPER ADORABLE :smiling_face_with_three_hearts:

Very attractive features:
- [X] Tasks management (todo, deadlines, events)
- [X] Tasks search
- [x] And more coming soon ...

so, WHAT ARE YOU WAITING FOR ?! :raised_eyebrow: :raised_eyebrow: :raised_eyebrow:

1. [Download it here!](https://github.com/laihuiqi/ip/releases/tag/A-Jar)
2. Run it :computer_mouse:
3. Have a nice chat
4. :tada: TaDaaa ! Get your task plan done! 
---
Want to do some fancy things :thinking:? You would like to play around with our bot's `Launcher` !!
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

laihuiqi committed:
change Duke.java to Chattime.java
complete ip week2 Level-1

laihuiqi committed:
add Week 2 Level 2

laihuiqi committed:
add class Task

laihuiqi committed:
add Week 2 Level 3

laihuiqi committed:
add Deadline.java, Event.java, Todo.java all inherit from Task class

laihuiqi committed:
add Week 2 Level 4

laihuiqi committed:
done TextUITesting, change Duke in runtest.sh to Chattime, add input.txt and EXPECTED.txt as test input and output

laihuiqi committed:
add ChattimeException class inherited from Exception

laihuiqi committed:
add Week 2 Level-5

laihuiqi committed:
no message

laihuiqi committed:
add Week 2 Level 6, use ArrayList to store tasks

laihuiqi committed:
modify test input, not implementing enum (A-Enums) because not necessary (based on current implementation of code

laihuiqi committed:
Add Level-7 Save in new branch.

laihuiqi committed:
Add Level 8 features. Add listTime command to enable date-task search.

laihuiqi committed:
Add gradle to project.

laihuiqi committed:
Implement OOP structure for the entire project.

laihuiqi committed:
Add package chattime with child packages command, exception, parser, storage, task and ui.

laihuiqi committed:
Add Level-9 find function.

laihuiqi committed:
no message

laihuiqi committed:
Add JavaDoc to project files.

laihuiqi committed:
Merge remote-tracking branch 'origin/add-gradle-support'. Resolve conflicts.

# Conflicts:
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

laihuiqi committed:
Add gradle and checkstyle automation.

laihuiqi committed:
Add JUnit test to all public non-void, non-static methods in project.

laihuiqi committed:
Package into jar file and set as 'pre-release'.

laihuiqi committed:
Check coding standards and conventions.
Ignore Javadoc for future branch merging.

laihuiqi committed:
Merge branch 'branch-A-JavaDoc' to 'master' branch.
Integrate JavaDoc into current branch.

# Conflicts:
#	src/main/java/chattime/Chattime.java
#	src/main/java/chattime/TaskList.java
#	src/main/java/chattime/command/AddCommand.java
#	src/main/java/chattime/command/ByeCommand.java
#	src/main/java/chattime/command/Command.java
#	src/main/java/chattime/command/DeleteCommand.java
#	src/main/java/chattime/command/ListCommand.java
#	src/main/java/chattime/command/MarkCommand.java
#	src/main/java/chattime/exception/ChattimeException.java
#	src/main/java/chattime/parser/Parser.java
#	src/main/java/chattime/storage/Storage.java
#	src/main/java/chattime/task/Deadline.java
#	src/main/java/chattime/task/Event.java
#	src/main/java/chattime/task/Task.java
#	src/main/java/chattime/task/Todo.java
#	src/main/java/chattime/ui/Ui.java

laihuiqi committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	.gitignore
#	src/main/java/chattime/Chattime.java
#	src/main/java/chattime/TaskList.java
#	src/main/java/chattime/command/AddCommand.java
#	src/main/java/chattime/command/ByeCommand.java
#	src/main/java/chattime/command/Command.java
#	src/main/java/chattime/command/DeleteCommand.java
#	src/main/java/chattime/command/ListCommand.java
#	src/main/java/chattime/command/MarkCommand.java
#	src/main/java/chattime/parser/Parser.java
#	src/main/java/chattime/task/Task.java
#	src/main/java/chattime/ui/Ui.java

laihuiqi committed:
no message

laihuiqi committed:
Revert "no message"

This reverts commit fa48523d52eed8baa00aa7da1a752db41030156c.

laihuiqi committed:
Modify all access modifiers in Parser.java to be static.
Delete constructor for Parser, create a parseInput method instead to
process the user input.

laihuiqi committed:
Add CheckStyle to chattime bot.

laihuiqi committed:
Add enum TaskType to Task.java (use in /chattime/task package). Modify
Deadline class constructor to take time input as varargs.

Tempura-Person commented on src/main/java/chattime/Chattime.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * A chatbot that receives user's input on various predetermined command types and performs relevant functions.
+ */
+public class Chattime {
```
I like the new name of the chatbot!

laihuiqi replied:
I'm so glad that you like my chatbot's name! Nice to meet you :smile:

Tempura-Person commented on src/main/java/chattime/command/AddCommand.java:
```diff
@@ -0,0 +1,39 @@
+        taskList.addTask(task);
+        storage.saveToFile(task);
+        ui.printAddTask(task, Task.printTotalTask());
+    }
```
I like how the AddCommand is used for all types of Tasks, too.

laihuiqi replied:
Thanks~

Tempura-Person commented on src/main/java/chattime/parser/Parser.java:
```diff
@@ -0,0 +1,315 @@
+    private static final String NO_INDEX = "OOPS!!! The index to %1$s cannot be empty.";
+    private static final String NEED_INT = "OOPS!!! The index to %1$s must be positive integer.";
+    private static final String IDX_OUT_OF_BOUND = "OOPS!!! The index is too large! We currently have %d task(s).";
+    private static final String UNRECOGNISED_COMMAND = "Sorry... but I don't understand what you said >,<";
```
I like the error messages being set as enums, very organised.

laihuiqi replied:
Yea, I'm finding a way to make things neat :wink:

Tempura-Person commented on src/main/java/chattime/parser/Parser.java:
```diff
@@ -0,0 +1,315 @@
+     *
+     * @throws ChattimeException Returns error message to require new input with problem statement.
+     */
+    private static void checkCleanCommand() throws ChattimeException {
```
Very nice, checking for this edge case.

laihuiqi replied:
Thank you for your comment !:star_struck:

Tempura-Person reviewed (commented):
Overall very impressive, organised, readable code that has sufficiently abstracted Classes.
All methods are well documented and well tested.
No complaints.

laihuiqi committed:
Add GUI to chat bot. Add `help` command to output user guide.

laihuiqi committed:
Fix minor issues, set window to be not resizable.

joshua-tyh commented on src/main/java/chattime/TaskList.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param index Input index from user, index-th task in task list to be removed.
+     */
+    public void removeListMember(int index) {
```
As the ArrayList has been defined to contain objects of type Task, maybe consider changing this method name to removeTask instead?

laihuiqi replied:
Thanks for your suggestion, it looks better now!

joshua-tyh commented on src/main/java/chattime/command/MarkCommand.java:
```diff
@@ -0,0 +1,50 @@
+     * Gets task from task list.
+     * Executes task status marking and returns UI message respectively to user.
+     * Updates data in storage file.
+     *  @param ui UI instance of bot.
```
Consider leaving a single line separator at line 31 and removing the extra space precessing @param to ensure consistency of Javadoc formats across code files.

laihuiqi replied:
Thanks for your reminder.

joshua-tyh commented on src/main/java/chattime/ui/Ui.java:
```diff
@@ -0,0 +1,139 @@
+                        + "   \\:\\ \\*H*A*T/ /  \\/__/*I*M*E\n"
+                        + "    \\:\\_\\    / :/  /\n"
+                        + "      \\/_/   \\/_/\n";
+
```
Very impressive logo!

laihuiqi replied:
Thank you!

joshua-tyh reviewed (commented):
Overall the code logic appears to be extremely thorough with great use of OOP. Just a few extremely minor tweaks to consider implementing. Keep up the great work! :)

laihuiqi committed:
Fix minor checkstyle problems.

laihuiqi committed:
`.chattime.storage.Storage ` `loadData()`: need checking on loaded data
                                           length
`.chattime.Chattime` `getResponse(String input)`: need checking on
                                                  processed commands

Lengths of data are specified for different types of tasks.
Processed commands should not be null.

Let's add assertions on lengths of loaded data according to types.
Let's add assertions that cmd should not be null.

The lengths of loaded data and the parsed commands are not accessible
by users, they need to have assertions to ensure that the programs are
running in a normal and good condition.

Assertion error messages are added for clarification.

laihuiqi committed:
`FindCommand.execute` method : long method
`Parser.parse` method : long method
`Storage.loadData` method : long method
`ChattimeException` class : can have better application of OOP

The 3 methods have long methods with LoC > 30, which may reduce the
readability.
The exception class can be use to have a better OOP utilized on
`toString()`.

Let's add `findIn` method to implement the find algorithm for
`FindCommand.execute`.
Let's add `parseAddCommand` and `parseIndexCommand` to `Parser` class
to lightweight `parse` method according to the cases functionality.
Let's add `createTaskFromStorage` method to create different types of
tasks based on the task code in the stored data.
Let's modify the `toString()` method to have a better generalized format
of error message.

Using sub- private functions can help to split the codes and enhance
readability.
Generalizing error messages output style can make the code short and
simple.

Add other minor changes to increase code quality.

laihuiqi committed:
`FindCommand.execute`: nested for-loop

`FindCommand.execute` contains nested for-loop that can be modified
using Streams.

Let's replace the for-loop with Streams.

Structure becomes more compact by using Streams.

Supplier is used to regenerate the Streams.

laihuiqi committed:
Merge pull request #1 from laihuiqi/branch-A-Assertions

A-Assertions Branch PR

laihuiqi committed:
Merge branch 'master' into branch-A-CodeQuality

laihuiqi committed:
Merge pull request #2 from laihuiqi/branch-A-CodeQuality

A-CodeQuality Branch PR

laihuiqi committed:
Merge branch 'master' into branch-A-Streams

laihuiqi committed:
Merge pull request #3 from laihuiqi/branch-A-Streams

A-Streams Branch PR

laihuiqi committed:
Add `ScheduleCommand` class to show daily schedule in hourly manner.

laihuiqi committed:
Modify files to suit the need of `schedule` command.

laihuiqi committed:
Add gradle.yml

laihuiqi committed:
Modify gradle.yml to remove cli output check

laihuiqi committed:
Fix minor issues.

laihuiqi committed:
Update execution permission for gradlew.

laihuiqi committed:
Modify `gradle.build` to suit jar file.
Add `isDuplicates` method in `TaskList` to detect duplicated input.

laihuiqi committed:
Modify the GUI to make window resizable, change the window title bar
icon, add names under profiles pictures.

laihuiqi committed:
Modify some bot-replies to give personality to Chattime.

laihuiqi committed:
Modify error handling method of `Storage` to reflect error message as
bot reply.
Modify time constraints for event, start <= end.
Improve coding quality, modify long method.

laihuiqi committed:
Add tests to original Junit test files.
Modify code according to checkStyle.

laihuiqi committed:
Add user guide in `docs/README.md` to help user familiar with Chattime.
Clarify commands through instructional images attached.
Modify `README.md` to suit Chattime's personality.

laihuiqi committed:
Change images. All of them are of own source.

laihuiqi committed:
Modify exit `bye` method, user can exit immediately by inputting `bye`.

laihuiqi committed:
Minor modification due to checkstyle.

laihuiqi committed:
Update README.md

laihuiqi committed:
Update README.md

laihuiqi committed:
Update user guide page theme.

laihuiqi committed:
Adjusts code to fix minor violations of javadoc convention.

laihuiqi committed:
Update `README.md` in docs.

laihuiqi committed:
Adjust README.md

laihuiqi committed:
Update README.md

laihuiqi committed:
Fix Linux pipeline problem with javafx version.
